torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+
+    size_t chunks_to_load = chunk_reader_.chunk_count();
+    chunk_sampler_.reset(chunks_to_load);
+
+    // Throw out any existing cached batch in the buffer and re-creates a new
+    // chunk buffer.
+    batch_buffer_ = torch::make_unique<
+        detail::BatchDataBuffer<UnwrappedBatchType, ExampleSamplerType>>(
+        chunks_to_load,
+        options_.batch_size_,
+        example_sampler_,
+        options_.ignore_empty_chunk_,
+        options_.cache_size_);
+
+    // create new workers for this new epoch.
+    quit_worker_ = false;
+
+    for (size_t i = 0; i < options_.preloader_count_; ++i) {
+      preload_threads_.emplace_back(
+          [this, i]() { this->preloader(i); });
+    }
+  }
+
+  /// size is not used for chunk dataset.
+  optional<size_t> size() const override {
+    return torch::nullopt;
+  }
+
+ private:
+  /// running on worker thread to preload chunk data.
+  void preloader(size_t id) {
+    while (!quit_worker_) {
+      try {
+        size_t chunk_id = 0;
+        if (auto chunk_sampler_result = chunk_sampler_.next(1)) {
+          chunk_id = chunk_sampler_result.value()[0];
+        } else {
+          break;
+        }
+        UnwrappedBatchType data = chunk_reader_.read_chunk(chunk_id);
+        AT_CHECK(options_.ignore_empty_chunk_ || !data.empty(),
+                 "Chunk with index " + std::to_string(chunk_id) + " is empty");
+
+        if (data.empty()) {
+          // skip adding the current chunk data and move to the next.
+          batch_buffer_->skip_chunk();
+        }
+        else {
+          batch_buffer_->add_chunk_data(std::move(data));
+        }
+      } catch (...) {
+        batch_buffer_->add_chunk_data(std::current_exception());
+      }
+    }
+  }
+
+  /// Block the current thread until the workers finish execution and exit.
+  void free_workers() {
+    if (!quit_worker_) {
+      quit_worker_ = true;
+      if(batch_buffer_){
+        batch_buffer_->stop();
+      }
+      for (auto& worker_thread : preload_threads_) {
+        worker_thread.join();
+      }
+    }
+  }
+
+ private:
+  // Templated class that defines what is a chunk and how to read chunk data.
+  // When a chunk is returned by chunk_reader_, ChunkDtatset split it into
+  // batches and caches them in batch_buffer_.
+  ChunkReader chunk_reader_;
+
+  // chunk sampler to shuffle different chunks
+  samplers::ThreadSafeSampler<ChunkSamplerType> chunk_sampler_;
+
+  // example sampler to shuffle examples in a specific chunk
+  ExampleSamplerType example_sampler_;
+
+  // batch data buffer which holds chunk data from preloading thread.
+  std::shared_ptr<detail::BatchDataBuffer<UnwrappedBatchType, ExampleSamplerType>>
+      batch_buffer_;
+
+  // worker thread pool
+  std::vector<std::thread> preload_threads_;
+
+  /// The options the Dataset was configured with.
+  const ChunkDatasetOptions options_;
+
+  // indicate whether the worker thread can be teared down
+  bool quit_worker_ = false;
```

Comment:
This is accessed from multiple threads without locking, so it should at least be an atomic value.

Reply:
It is only a private member and used by the reset method, which is not thread safe. So not sure what we achieve by making this atomic.

Reply:
This variable is always set/write from the main thread and read from the worker threads, so we don't have a race writer here. Though it is definitely a multi-thread scenario, when it comes the question of do I want to add synchronization cost, I ask if I do care the reader gets stale value. In this case, stale quit_worker value doesn't bother with proper functionality and it will get refreshed in next loop. So, it is unnecessary to make it atomic with undesired sync cost.

Reply:
This is irrelevant: if you write from one thread and read from another thread without an intervening synchronization, you have a race. For example, if a compiler can prove that a store is dead within a single threaded context, it is allowed to remove the store, even if there is a read from another thread that is relying on the store happening.

If we are really concerned about the cost of atomic (which may be warranted), we should specify a specific `std::memory_order`. On a TSO machine those operations will compile into normal, non-LOCK'ing instructions. But if you are feeling too lazy to reason about what the correct memory order is in this situation, you should use an atomic.

Reply:
>if you write from one thread and read from another thread without an intervening synchronization, you have a race

Not really. Firstly there is only one writer thread, which means there is no writer racing. Secondly it is a Boolean type, so cases like part of the value is updated, part of it is not doesn't happen. Now it is possible that the reader may read stale (old) value from the reader thread. But as I mentioned earlier, this doesn't matter here as it will get refreshed in the next check. As that being said, I think no synchronization is needed in this scenario.

Reply:
You are right this actually could be problematic. Though I don't think the compiler will optimize out member variable quit_worker_, it is still possible that the old value is cached in worker thread such that the new value is never used. The actually behavior is up to different compiler, but we should not risk it.
Fixed it by using atomic. Thanks!